### PR TITLE
Add default impl to handle token challenges

### PR DESCRIPTION
**Pull request by xiangyan99**

### CODE REVIEW

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py [hunk 1]
```diff
@@ -191,6 +195,21 @@ def on_challenge(
...
+        if error == "insufficient_claims":
+            encoded_claims = get_challenge_parameter(headers, "Bearer", "claims")
+            if not encoded_claims:
+                return False
+            claims = base64.urlsafe_b64decode(encoded_claims).decode("utf-8")
+            if claims:
+                try:
+                    token = self._get_token(*self._scopes, claims=claims)
+                    bearer_token = cast(Union["AccessToken", "AccessTokenInfo"], token).token
+                    self._update_headers(request.http_request.headers, bearer_token)
```
Comment (pvaneck) [inline sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:209]: I see we are not caching the token retrieved in this case (i.e. using `self.authorize_request(request, *self._scopes, claims=claims)`). Is this intentional, and we expect another get_token call on subsequent requests?

File: sdk/core/azure-core/azure/core/pipeline/policies/_utils.py [hunk 2]
```diff
@@ -102,3 +102,101 @@ def get_domain(url: str) -> str:
...
+    scheme = challenge_scheme
+    parameter = challenge_parameter
+    header_span = header_value
+
+    # Iterate through each challenge value.
+    while get_next_challenge(header_span):
+        challenge = get_next_challenge(header_span)
+        if not challenge:
+            break
+        challenge_key, header_span = challenge
```
Comment (pvaneck) [inline sdk/core/azure-core/azure/core/pipeline/policies/_utils.py:130]: Can we add the following to save on the next loop if the challenge_key isn't what we are looking for:

```python
        if challenge_key.lower() != scheme.lower():
            continue
```

**Review by pvaneck — COMMENTED** [review]
Looking pretty good. For posterity, a blurb in the PR description about the motivation for this would be helpful. Left a few questions and comments in-line.

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py [hunk 3]
```diff
@@ -191,6 +195,21 @@ def on_challenge(
         :rtype: bool
         """
         # pylint:disable=unused-argument
+        headers = response.http_response.headers
+        error = get_challenge_parameter(headers, "Bearer", "error")
+        if error == "insufficient_claims":
+            encoded_claims = get_challenge_parameter(headers, "Bearer", "claims")
+            if not encoded_claims:
+                return False
+            claims = base64.urlsafe_b64decode(encoded_claims).decode("utf-8")
```
[the remark this answers is not in the document]
Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:204]: This is the implementation of the .net code. Chris told me there was no service on board yet (key vault uses CAE but it has its own implementation). Services will use MISE to enable CAE and my understanding is that this is the behavior of MISE. If services do not follow this, they may need their own implementation in the SDKs.

[the remark this answers is not in the document]
Comment (pvaneck) [reply sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:204]: I think to be a bit on the safer side, we should maintain the additional padding logic and have the decode method in the try/except block, so the app doesn't crash if the claims string is somehow malformed. Some claims could at some point in the future be encoded as URL-safe Base64, and padding could be omitted, leading to "incorrect padding" errors in the decode method. Cheap safeguard to add.
```python
            try:
                padding_needed = -len(encoded_claims) % 4
                claims = base64.urlsafe_b64decode(encoded_claims + "=" * padding_needed).decode("utf-8")
                if claims:
                    token = await self._get_token(*self._scopes, claims=claims)
                    bearer_token = cast(Union["AccessToken", "AccessTokenInfo"], token).token
                    request.http_request.headers["Authorization"] = "Bearer " + bearer_token
                    return True
            except Exception:  # pylint:disable=broad-except
                return False
```

**Review by pvaneck — APPROVED** [review]

**Pull request merged**